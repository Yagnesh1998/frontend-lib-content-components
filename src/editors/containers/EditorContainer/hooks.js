import { useSelector } from 'react-redux';

import { RequestKeys } from '../../data/constants/requests';
import { selectors } from '../../data/redux';
import * as appHooks from '../../hooks';
import analyticsEvt from '../../data/constants/analyticsEvt';

export const {
  navigateCallback,
  nullMethod,
  saveBlock,
} = appHooks;

export const setAssetToStaticUrl = (images, getContent) => {
  let content = getContent();
  const imageUrls = [];
  const imgsArray = Object.values(images);
  imgsArray.forEach(image => {
    imageUrls.push({ portableUrl: image.portableUrl, displayName: image.displayName });
  });
  const imageSrcs = content.split('src="');
  imageSrcs.forEach(src => {
    if (src.startsWith('/asset') && imageUrls.length > 0) {
      const nameFromEditorSrc = src.substring(src.lastIndexOf('@') + 1, src.indexOf('"'));
      const nameFromStudioSrc = nameFromEditorSrc.substring(nameFromEditorSrc.indexOf('/') + 1);
      let portableUrl;
      imageUrls.forEach((url) => {
        if (url.displayName === nameFromEditorSrc || url.displayName === nameFromStudioSrc) {
          portableUrl = url.portableUrl;
        }
      });
      if (portableUrl) {
        const currentSrc = src.substring(0, src.indexOf('"'));
        const updatedContent = content.replace(currentSrc, portableUrl);
        content = updatedContent;
      }
    }
  });
  return content;
};

export const handleSaveClicked = ({ getContent, dispatch }) => {
  const destination = useSelector(selectors.app.returnUrl);
  const analytics = useSelector(selectors.app.analytics);
  const images = useSelector(selectors.app.images);
  return () => saveBlock({
    content: setAssetToStaticUrl(images, getContent),
    destination,
    analytics,
    dispatch,
  });
};
export const handleCancelClicked = ({ onClose }) => {
  if (onClose) {
    return onClose;
  }
  return navigateCallback({
    destination: useSelector(selectors.app.returnUrl),
    analyticsEvent: analyticsEvt.editorCancelClick,
    analytics: useSelector(selectors.app.analytics),
  });
};
export const isInitialized = () => useSelector(selectors.app.isInitialized);
export const saveFailed = () => useSelector((state) => (
  selectors.requests.isFailed(state, { requestKey: RequestKeys.saveBlock })
));
