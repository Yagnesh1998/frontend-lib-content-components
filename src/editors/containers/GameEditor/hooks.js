import { useState, useEffect } from 'react';
import { Form } from '@edx/paragon';
import './index.scss';
