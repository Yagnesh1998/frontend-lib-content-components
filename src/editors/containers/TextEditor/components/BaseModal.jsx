import React from 'react';
import PropTypes from 'prop-types';

import {
  ActionRow,
  ModalDialog,
} from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import messages from './messages';

export const BaseModal = ({
  isOpen,
  close,
  title,
  children,
  confirmAction,
  footerAction,
  size,
}) => (
  <ModalDialog
    isOpen={isOpen}
    onClose={close}
    size={size}
    variant="default"
    hasCloseButton
    isFullscreenOnMobile
    isFullscreenScroll
  >
    <ModalDialog.Header>
      <ModalDialog.Title>
        {title}
      </ModalDialog.Title>
    </ModalDialog.Header>
    <ModalDialog.Body>
      {children}
    </ModalDialog.Body>
    <ModalDialog.Footer>
      <ActionRow>
        {footerAction}
        <ActionRow.Spacer />
        <ModalDialog.CloseButton variant="tertiary" onClick={close}>
          <FormattedMessage {...messages.cancelButtonLabel} />
        </ModalDialog.CloseButton>
        {confirmAction}
      </ActionRow>
    </ModalDialog.Footer>
  </ModalDialog>
);

BaseModal.defaultProps = {
  footerAction: null,
  size: 'lg',
};

BaseModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  close: PropTypes.func.isRequired,
  title: PropTypes.node.isRequired,
  children: PropTypes.node.isRequired,
  confirmAction: PropTypes.node.isRequired,
  footerAction: PropTypes.node,
  size: PropTypes.string,
};

export default BaseModal;
