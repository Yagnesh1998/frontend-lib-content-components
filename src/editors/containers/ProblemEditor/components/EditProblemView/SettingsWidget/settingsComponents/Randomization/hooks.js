import { useState, useEffect } from 'react';
import messages from './messages';
import { RandomizationTypes } from '../../../../../../../data/constants/problem';
import * as module from './hooks';

export const state = {
  summary: (val) => useState(val),
};

export const useRandomizationSettingStatus = ({ randomization, updateSettings }) => {
  const [summary, setSummary] = module.state.summary({ message: messages.noRandomizationSummary, values: {} });
  useEffect(() => {
    setSummary({
      message: randomization ? RandomizationTypes[randomization] : messages.noRandomizationSummary,
    });
  }, [randomization]);

  const handleChange = (event) => {
    updateSettings({ randomization: event.target.value });
  };
  return { summary, handleChange };
};

export default useRandomizationSettingStatus;
