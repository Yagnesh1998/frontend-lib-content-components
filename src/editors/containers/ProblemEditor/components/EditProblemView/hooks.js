import 'tinymce';
import ReactStateSettingsParser from '../../data/ReactStateSettingsParser';
import ReactStateOLXParser from '../../data/ReactStateOLXParser';
import { setAssetToStaticUrl } from '../../../../sharedComponents/TinyMceWidget/hooks';

export const fetchEditorContent = () => {
  const editorObject = {};
  const EditorsArray = window.tinymce.editors;
  Object.entries(EditorsArray).forEach(([id, editor]) => {
    if (Number.isNaN(parseInt(id))) {
      if (id.startsWith('answer')) {
        const { answers } = editorObject;
        const answerId = id.substring(id.indexOf('-') + 1);
        editorObject.answers = { ...answers, [answerId]: editor.getContent() };
      } else {
        editorObject[id] = editor.getContent();
      }
    }
  });
  return editorObject;
};

export const parseState = ({
  problem,
  isAdvanced,
  ref,
  assets,
  lmsEndpointUrl,
}) => () => {
  const editorObject = fetchEditorContent();
  const reactSettingsParser = new ReactStateSettingsParser(problem);
  const reactOLXParser = new ReactStateOLXParser({ problem, editorObject });
  const reactBuiltOlx = setAssetToStaticUrl({ editorValue: reactOLXParser.buildOLX(), assets, lmsEndpointUrl });
  const rawOLX = ref?.current?.state.doc.toString();
  return {
    settings: reactSettingsParser.getSettings(),
    olx: isAdvanced ? rawOLX : reactBuiltOlx,
  };
};
