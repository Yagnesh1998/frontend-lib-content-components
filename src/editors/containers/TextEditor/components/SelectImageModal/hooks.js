import React from 'react';
import { useDispatch } from 'react-redux';

import { thunkActions } from '../../../../data/redux';
import * as module from './hooks';
import { sortFunctions, sortKeys } from './utils';

export const state = {
  highlighted: (val) => React.useState(val),
  images: (val) => React.useState(val),
  showSelectImageError: (val) => React.useState(val),
  searchString: (val) => React.useState(val),
  sortBy: (val) => React.useState(val),
  showSizeError: (val) => React.useState(val),
};

export const searchAndSortHooks = () => {
  const [searchString, setSearchString] = module.state.searchString('');
  const [sortBy, setSortBy] = module.state.sortBy(sortKeys.dateNewest);
  return {
    searchString,
    onSearchChange: (e) => setSearchString(e.target.value),
    clearSearchString: () => setSearchString(''),
    sortBy,
    onSortClick: (key) => () => setSortBy(key),
  };
};

export const filteredList = ({ searchString, imageList }) => (
  imageList.filter(({ displayName }) =>displayName.toLowerCase().includes(searchString.toLowerCase()))
);

export const displayList = ({ sortBy, searchString, images }) =>
  module.filteredList({
      searchString,
      imageList: Object.values(images),
    }).sort(sortFunctions[sortBy in sortKeys ? sortKeys[sortBy] : sortKeys.dateNewest]);

export const imgListHooks = ({ searchSortProps, setSelection }) => {
  const dispatch = useDispatch();
  const [images, setImages] = module.state.images({});
  const [highlighted, setHighlighted] = module.state.highlighted(null);
  const [
    showSelectImageError,
    setShowSelectImageError,
  ] = module.state.showSelectImageError(false);
  const [showSizeError, setShowSizeError] = module.state.showSelectImageError(false);
  const list = module.displayList({ ...searchSortProps, images });

  React.useEffect(() => {
    dispatch(thunkActions.app.fetchImages({ setImages }));
  }, []);

  return {
    galleryError: {
      show: showSelectImageError,
      set: () => setShowSelectImageError(true),
      dismiss: () => setShowSelectImageError(false),
    },
    inputError: {
      show: showSizeError,
      set: () => setShowSizeError(true),
      dismiss: () => setShowSizeError(false),
    },
    images,
    galleryProps: {
      galleryIsEmpty: Object.keys(images).length === 0,
      searchIsEmpty: list.length === 0,
      displayList: list,
      highlighted,
      onHighlightChange: (e) => setHighlighted(e.target.value),
    },
    // highlight by id
    selectBtnProps: {
      onClick: () => {
        if (highlighted) {
          setSelection(images[highlighted]);
        } else {
          setShowSelectImageError(true);
        }
      },
    },
  };
};

export const checkValidFileSize = ({
  selectedFile,
  clearSelection,
  onSizeFail,
}) => {
  const fileSize = (selectedFile.size / 1000000).toFixed(4);
  if (fileSize > 10) {
    clearSelection();
    onSizeFail();
    return false;
  }
  return true;
};

export const fileInputHooks = ({ setSelection, clearSelection, imgList }) => {
  const dispatch = useDispatch();
  const ref = React.useRef();
  const click = () => ref.current.click();
  const addFile = (e) => {
    const selectedFile = e.target.files[0];
    if (selectedFile && module.checkValidFileSize({
        selectedFile,
        clearSelection,
        onSizeFail: () => {
          imgList.inputError.set();
        },
      })
    ) {
      dispatch(
        thunkActions.app.uploadImage({
          file: selectedFile,
          setSelection,
        }),
      );
    }
  };

  return {
    click,
    addFile,
    ref,
  };
};

export const imgHooks = ({ setSelection, clearSelection }) => {
  const searchSortProps = module.searchAndSortHooks();
  const imgList = module.imgListHooks({ setSelection, searchSortProps });
  const fileInput = module.fileInputHooks({
    setSelection,
    clearSelection,
    imgList,
  });
  const {
    galleryError,
    galleryProps,
    inputError,
    selectBtnProps,
  } = imgList;

  return {
    galleryError,
    inputError,
    fileInput,
    galleryProps,
    searchSortProps,
    selectBtnProps,
  };
};

export default {
  imgHooks,
};
