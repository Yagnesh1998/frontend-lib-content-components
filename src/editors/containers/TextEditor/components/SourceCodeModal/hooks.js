import { useRef } from 'react';
import * as module from './hooks';

export const getSaveBtnProps = ({ editorRef, ref, close }) => ({
  onClick: () => {
    if (editorRef && editorRef.current && ref && ref.current) {
      const content = ref.current.state.doc.toString();
      editorRef.current.setContent(content);
      close();
    }
  },
});

export const preprareSourceCodeModal = ({ editorRef, close }) => {
  const ref = useRef();
  const saveBtnProps = module.getSaveBtnProps({ editorRef, ref, close });
  const value = editorRef.current?.getContent();
  return { saveBtnProps, value, ref };
};

export default {
  preprareSourceCodeModal,
};
