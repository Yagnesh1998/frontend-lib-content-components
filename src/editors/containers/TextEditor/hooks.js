import {
  useRef, useEffect, useCallback, useState,
} from 'react';

import { StrictDict } from '../../utils';
import tinyMCE from '../../data/constants/tinyMCE';
import pluginConfig from './pluginConfig';
import * as appHooks from '../../hooks';
import * as module from './hooks';

export const { nullMethod, navigateCallback, navigateTo } = appHooks;

export const state = StrictDict({
  isModalOpen: (val) => useState(val),
  imageSelection: (val) => useState(val),
  refReady: (val) => useState(val),
});

export const setupCustomBehavior = ({ openModal, setImage }) => (editor) => {
  // image upload button
  editor.ui.registry.addButton(tinyMCE.buttons.imageUploadButton, {
    icon: 'image',
    tooltip: 'Add Image',
    onAction: openModal,
  });
  // editing an existing image
  editor.ui.registry.addButton(tinyMCE.buttons.editImageSettings, {
    icon: 'image',
    tooltip: 'Edit Image Settings',
    onAction: module.openModalWithSelectedImage({ editor, setImage, openModal }),
  });
  // overriding the code plugin's icon with 'HTML' text
  const openCodeEditor = () => editor.execCommand('mceCodeEditor');
  editor.ui.registry.addButton(tinyMCE.buttons.code, {
    text: 'HTML',
    tooltip: 'Source code',
    onAction: openCodeEditor,
  });
  // add a custom simple inline code block formatter.
  const toggleCodeFormatting = () => editor.formatter.toggle('code');
  editor.ui.registry.addButton(tinyMCE.buttons.codeBlock, {
    icon: 'sourcecode',
    tooltip: 'Code Block',
    onAction: toggleCodeFormatting,
  });
  const toggleBlockQuoteFormatting = () => editor.formatter.toggle('blockquote');
  editor.ui.registry.addButton(tinyMCE.buttons.blockQuote, {
    icon: 'quote',
    tooltip: 'Block Quote',
    onAction: toggleBlockQuoteFormatting,
  });
};

// imagetools_cors_hosts needs a protocol-sanatized url
export const removeProtocolFromUrl = (url) => url.replace(/^https?:\/\//, '');

export const editorConfig = ({
  setEditorRef,
  blockValue,
  openModal,
  initializeEditor,
  setSelection,
  lmsEndpointUrl,
  studioEndpointUrl,
}) => ({
  onInit: (evt, editor) => {
    setEditorRef(editor);
    initializeEditor();
  },
  initialValue: blockValue ? blockValue.data.data : '',
  init: {
    content_style: "@import url('https://fonts.googleapis.com/css2?family=Open+Sans'); body { color: #313131; font-family: 'Open Sans'; font-size: 16px;",
    setup: module.setupCustomBehavior({ openModal, setImage: setSelection }),
    plugins: pluginConfig.plugins,
    imagetools_toolbar: pluginConfig.imageToolbar,
    toolbar: pluginConfig.toolbar,
    contextmenu: 'link table',
    ...pluginConfig.config,
    valid_elements: '*[*]',
    valid_children: '+body[style]',
    imagetools_cors_hosts: [removeProtocolFromUrl(lmsEndpointUrl), removeProtocolFromUrl(studioEndpointUrl)],
  },
});

export const modalToggle = () => {
  const [isOpen, setIsOpen] = module.state.isModalOpen(false);
  return {
    isOpen,
    openModal: () => setIsOpen(true),
    closeModal: () => setIsOpen(false),
  };
};

export const openModalWithSelectedImage = ({ editor, setImage, openModal }) => () => {
  const imgHTML = editor.selection.getNode();
  setImage({
    externalUrl: imgHTML.src,
    altText: imgHTML.alt,
    width: imgHTML.width,
    height: imgHTML.height,
  });
  openModal();
};

export const prepareEditorRef = () => {
  const editorRef = useRef(null);
  const setEditorRef = useCallback((ref) => {
    editorRef.current = ref;
  }, []);
  const [refReady, setRefReady] = module.state.refReady(false);
  useEffect(() => setRefReady(true), []);
  return { editorRef, refReady, setEditorRef };
};

export const getContent = ({ editorRef, isRaw }) => () => {
  if (isRaw && editorRef && editorRef.current) {
    return editorRef.current.value;
  }
  return editorRef.current?.getContent();
};

export const selectedImage = (val) => {
  const [selection, setSelection] = module.state.imageSelection(val);
  return {
    clearSelection: () => setSelection(null),
    selection,
    setSelection,
  };
};
