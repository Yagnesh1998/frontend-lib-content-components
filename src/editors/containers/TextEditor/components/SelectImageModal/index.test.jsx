import React from 'react';
import { shallow } from 'enzyme';

import { formatMessage } from '../../../../../testUtils';
import { RequestKeys } from '../../../../data/constants/requests';
import { selectors } from '../../../../data/redux';
import BaseModal from '../BaseModal';
import FileInput from '../../../../sharedComponents/FileInput';
import Gallery from './Gallery';
import SearchSort from './SearchSort';
import hooks from './hooks';
import { SelectImageModal, mapStateToProps, mapDispatchToProps } from '.';

jest.mock('../BaseModal', () => 'BaseModal');
jest.mock('../../../../sharedComponents/FileInput', () => 'FileInput');
jest.mock('./Gallery', () => 'Gallery');
jest.mock('./SearchSort', () => 'SearchSort');
jest.mock('../../../../sharedComponents/ErrorAlerts/FetchErrorAlert', () => 'FetchErrorAlert');
jest.mock('../../../../sharedComponents/ErrorAlerts/UploadErrorAlert', () => 'UploadErrorAlert');
jest.mock('../../../../sharedComponents/ErrorAlerts/ErrorAlert', () => 'ErrorAlert');

jest.mock('./hooks', () => ({
  imgHooks: jest.fn(() => ({
    galleryError: {
      show: 'ShoWERror gAlLery',
      set: jest.fn(),
      dismiss: jest.fn(),
    },
    inputError: {
      show: 'ShoWERror inPUT',
      set: jest.fn(),
      dismiss: jest.fn(),
    },
    fileInput: {
      addFile: 'imgHooks.fileInput.addFile',
      click: 'imgHooks.fileInput.click',
      ref: 'imgHooks.fileInput.ref',
    },
    galleryProps: { gallery: 'props' },
    searchSortProps: { search: 'sortProps' },
    selectBtnProps: { select: 'btnProps' },
  })),
}));

jest.mock('../../../../data/redux', () => ({
  selectors: {
    requests: {
      isPending: (state, { requestKey }) => ({ isPending: { state, requestKey } }),
    },
  },
}));

describe('SelectImageModal', () => {
  describe('component', () => {
    const props = {
      isOpen: true,
      close: jest.fn().mockName('props.close'),
      setSelection: jest.fn().mockName('props.setSelection'),
      clearSelection: jest.fn().mockName('props.clearSelection'),
      intl: { formatMessage },
      inputIsLoading: false,
    };
    let el;
    const imgHooks = hooks.imgHooks();
    beforeEach(() => {
      el = shallow(<SelectImageModal {...props} />);
    });
    test('snapshot', () => {
      expect(el).toMatchSnapshot();
    });
    test('snapshot: uploaded image not loaded, show spinner', () => {
      props.inputIsLoading = true;
      expect(shallow(<SelectImageModal {...props} />)).toMatchSnapshot();
      props.inputIsLoading = false;
    });
    it('provides confirm action, forwarding selectBtnProps from imgHooks', () => {
      expect(el.find(BaseModal).props().confirmAction.props).toEqual(
        expect.objectContaining({ ...hooks.imgHooks().selectBtnProps, variant: 'primary' }),
      );
    });
    it('provides file upload button linked to fileInput.click', () => {
      expect(el.find(BaseModal).props().footerAction.props.onClick).toEqual(
        imgHooks.fileInput.click,
      );
    });
    it('provides a SearchSort component with searchSortProps from imgHooks', () => {
      expect(el.find(SearchSort).props()).toEqual(imgHooks.searchSortProps);
    });
    it('provides a Gallery component with galleryProps from imgHooks', () => {
      expect(el.find(Gallery).props()).toEqual(imgHooks.galleryProps);
    });
    it('provides a FileInput component with fileInput props from imgHooks', () => {
      expect(el.find(FileInput).props()).toMatchObject({ fileInput: imgHooks.fileInput });
    });
  });
  describe('mapStateToProps', () => {
    const testState = { some: 'testState' };
    test('loads inputIsLoading from requests.isPending selector for uploadImage request', () => {
      expect(mapStateToProps(testState).inputIsLoading).toEqual(
        selectors.requests.isPending(testState, { requestKey: RequestKeys.uploadImage }),
      );
    });
  });
  describe('mapDispatchToProps', () => {
    test('is empty', () => {
      expect(mapDispatchToProps).toEqual({});
    });
  });
});
