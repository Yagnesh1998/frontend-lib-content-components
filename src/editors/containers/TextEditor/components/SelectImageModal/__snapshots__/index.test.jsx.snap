// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectImageModal component snapshot 1`] = `
<BaseModal
  close={[MockFunction props.close]}
  confirmAction={
    <Button
      select="btnProps"
      variant="primary"
    >
      <FormattedMessage
        defaultMessage="Next"
        description="Label for Next button"
        id="authoring.texteditor.selectimagemodal.next.label"
      />
    </Button>
  }
  footerAction={
    <Button
      onClick="imgHooks.fileInput.click"
      variant="link"
    >
      <FormattedMessage
        defaultMessage="Upload a new image (10 MB max)"
        description="Label for upload button"
        id="authoring.texteditor.selectimagemodal.upload.label"
      />
    </Button>
  }
  isOpen={true}
  title="Add an image"
>
  <FetchErrorAlert />
  <UploadErrorAlert />
  <ErrorAlert
    dismissError={[MockFunction]}
    hideHeading={true}
    isError="ShoWERror inPUT"
  >
    <FormattedMessage
      defaultMessage="Images must be 10 MB or less. Please resize image and try again."
      description=" Message presented to user when file size of image is larger than 10 MB"
      id="authoring.texteditor.selectimagemodal.error.fileSizeError"
    />
  </ErrorAlert>
  <ErrorAlert
    dismissError={[MockFunction]}
    hideHeading={true}
    isError="ShoWERror gAlLery"
  >
    <FormattedMessage
      defaultMessage="Select an image to continue."
      description="Message presented to user when clicking Next without selecting an image"
      id="authoring.texteditor.selectimagemodal.error.selectImageError"
    />
  </ErrorAlert>
  <Stack
    gap={3}
  >
    <SearchSort
      search="sortProps"
    />
    <Gallery
      gallery="props"
    />
    <FileInput
      fileInput={
        Object {
          "addFile": "imgHooks.fileInput.addFile",
          "click": "imgHooks.fileInput.click",
          "ref": "imgHooks.fileInput.ref",
        }
      }
    />
  </Stack>
</BaseModal>
`;
